crypto/groups/qr_rsa.go

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2017 XLAB d.o.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package groups
+
+import (
+	"errors"
+	"github.com/xlab-si/emmy/crypto/common"
+	"math/big"
+)
+
+// QRRSA presents group of quadratic residues modulo N where N is a product
+// of two primes. This group is in general NOT cyclic (it is only when (P-1)/2 and (Q-1)/2 are primes,
+// see QRSpecialRSA).
+type QRRSA struct {
+	N     *big.Int // N = P * Q
+	P     *big.Int
+	Q     *big.Int
+	Order *big.Int
+}
+
+func NewQRRSA(P, Q *big.Int) (*QRRSA, error) {
+	if !P.ProbablyPrime(20) || !Q.ProbablyPrime(20) {
+		return nil, errors.New("P and Q must be be primes")
```

Comment:
I think it is good to have such checks - it doesn't hurt (it makes it a little bit slower, but until performance is not an issue ... ) and it can prevent serious mistakes. Crypto API should not be built relying on user's expertise. Also, imagine you mistakenly change P and Q stored in config file.

Reply:
I don't imagine this function being the part of this API that is actually exposed to an end-user. IMO every function does not need to be designed like that, there's just too many things that can go wrong in using crypto APIs by novice users. The highest level functions should, of course, be designed to prevent any inproper use.

IMO it would be much better to include this kind of checks in the code that reads the config file.

I will leave this up to you though.

Reply:
Arguments from both of you make total sense to me.

But nevertheless, I agree with @miha-stopar that you can never be too careful. Also, in terms of what these checks are doing (checking whether the provided numbers are primes or not), I don't think they semantically belong to higher-level functions. E.g. do we want to bother higher-level functions with this (IMO) strictly crypto safeguard? 

For instance, in our case the `config.go` would suddenly have to be aware of these checks (checking primes). I don't think it's the right place for that :slightly_smiling_face:

Reply:
The higher level functions would probably not bother with the checks either, because they would be the ones calling functions to generate prime numbers.

I imagine having a function to generate the server's parameters, which would output the values appropriate for the config file. In case someone wants to be smart and change those, `config.go` seems just the perfect place to check whether the configuration makes sense.

This one check doensn't hurt indeed, but if we would stick to this practice, having this kind of checks in every function would probably hurt the performance.

Reply:
I want to have these checks. Crypto API should not make any assumptions. Not the same, but related: https://blog.cryptographyengineering.com/2012/12/28/the-anatomy-of-bad-idea/